[xh]Add new relic monitoring

matrixstone opened the pull request:
# Summary
Add new relic monitoring.

- Create a `ENABLE_NEW_RELIC` variable to enable or disable new relic monitoring.
- User need to follow new relic guide to create configuration file with license_key and app name.
- An example of new relic configuration file `example_newrelic.ini` is attached.
- A documentation of new relic integration is added.

# Tests

Error reporting for `python3 mage_ai/cli/main.py run project_name pipeline_name`
![screenshot1](https://github.com/mage-ai/mage-ai/assets/5386254/449705f4-e309-4a82-a07e-4abff2beb265)

Error reporting with triggers with worker run
![Monitoring from process queue](https://github.com/mage-ai/mage-ai/assets/5386254/a9c79b5d-21ac-4f7a-a44c-b00cc30a0393)

Logs from scheduler run
![logs_with_schedule](https://github.com/mage-ai/mage-ai/assets/5386254/6da7cc33-bbee-456d-9ead-84ddf28ce3b1)



cc:
@wangxiaoyou1993 @dy46

wangxiaoyou1993 commented on example_newrelic.ini:
```diff
@@ -0,0 +1,210 @@
+# ---------------------------------------------------------------------------
```
Can you put this file in `newrelic` subfolder under https://github.com/mage-ai/mage-ai/tree/master/integrations ?

wangxiaoyou1993 commented on docs/production/observability/newrelic.mdx:
```diff
@@ -0,0 +1,32 @@
+---
```
To add the doc to navigation bar, you'll need to add the doc to this file: https://github.com/mage-ai/mage-ai/blob/master/docs/mint.json#L409

matrixstone replied:
Done

wangxiaoyou1993 commented on mage_ai/server/scheduler_manager.py:
```diff
@@ -24,9 +26,14 @@ def run_scheduler():
-        database_manager.run_migrations()
+        with newrelic.agent.BackgroundTask(application, name="schedule-run", group='Task') \
+             if ENABLE_NEW_RELIC else nullcontext():
+            database_manager.run_migrations()
```
this is just db migration. it's not the scheduler process.

matrixstone replied:
Updated to name it as db-migration in new relic report

wangxiaoyou1993 commented on mage_ai/orchestration/queue/process_queue.py:
```diff
@@ -89,7 +89,10 @@ def __init__(self, queue: mp.Queue, job_dict):
                 traces_sample_rate=SENTRY_TRACES_SAMPLE_RATE,
             )
+        if ENABLE_NEW_RELIC:
+            newrelic.agent.initialize('/home/src/newrelic.ini')
```
same. let's not hardcode config path

wangxiaoyou1993 replied:
what happens if the newrelic config doesn't exist or is invalid, does the app still run correctly in that case?

matrixstone replied:
It fails if connfiguration missing. I now wrap it with a try catch clause to capture the configuration exception to make the code proceed

matrixstone commented:
Thanks for the review @wangxiaoyou1993! Addressed all comments. PTAL.

wangxiaoyou1993 commented on mage_ai/cli/main.py:
```diff
@@ -171,55 +179,66 @@ def run(
+        try:
+            newrelic.agent.initialize(NEW_RELIC_CONFIG_PATH)
+            application = newrelic.agent.register_application(timeout=10)
+        except newrelic.api.exceptions.ConfigurationError as error:
```
Is this the only possible exception? Could there be other exceptions?

matrixstone replied:
I added a general `except Exception` clause.

Even through there are only two exceptions defined by new relic python API: https://github.com/newrelic/newrelic-python-agent/blob/main/newrelic/api/exceptions.py#L16 It is better to have a general Exception catch here in case new exception introduced by newrelic in future.

wangxiaoyou1993 commented on mage_ai/server/scheduler_manager.py:
```diff
@@ -24,9 +32,20 @@ def run_scheduler():
             sentry_dsn,
             traces_sample_rate=SENTRY_TRACES_SAMPLE_RATE,
         )
+    disable_new_relic = False
```
Can we try reducing duplicate code?
You can add a newrelic subfolder in https://github.com/mage-ai/mage-ai/tree/master/mage_ai/services and put newrelic related methods in it so that you can reuse it.

matrixstone replied:
Yes, created a common function under `services/newrelic` and reused them in different scripts.

matrixstone commented:
Thanks for the review @wangxiaoyou1993! Addressed all comments.

Here are screenshots in new relic after the code change:

Log reported to new relic:
![log_exported1](https://github.com/mage-ai/mage-ai/assets/5386254/888943b1-db51-4cef-982c-a3c8c6081e05)

When there is error and the pipeline still complete, log printed:
![Error_example1](https://github.com/mage-ai/mage-ai/assets/5386254/094338f2-d869-4edb-a6e4-e74e3b66a3e1)
![tasks complete1](https://github.com/mage-ai/mage-ai/assets/5386254/fa57189d-9e6b-4ef3-81ad-793c73eca91d)
